tc/sgx/trusted_worker_manager/enclave/kme/workload/kme_workload_plug-in.cpp

```diff
@@ -57,6 +75,268 @@ WPEInfo::WPEInfo(const ByteArray& _sk) {
+
+    SetStatus(KME_REPL_OP_SUCCESS, out_work_order_data);
+    AddOutput(1, uid_ba, out_work_order_data);
+}
+
+/*
+ *  Create KME State Update Request.
+ *
+ *  @param in_work_order_data - vector of work order indata.
+ *         It consists of the unique_id and signing key
+ *  @param out_work_order_data - vector of work order outdata
+*/
+void KMEWorkloadProcessor::CreateStateReplicatonRequest(
+    const std::vector<tcf::WorkOrderData>& in_work_order_data,
+    std::vector<tcf::WorkOrderData>& out_work_order_data) {
+
+    // Extract uid from in_data that comes in the request from
+    // replica KME
+    ByteArray uid_ba = in_work_order_data[0].decrypted_data;
+    // Generate random bytes of size 16,
+    ByteArray nonce_bytes = tcf::crypto::RandomBitString(16);
+    // Convert nonce to hex string
+    std::string nonce_hex = ByteArrayToHexEncodedString(nonce_bytes);
+    // Persist the state_req_nonce_hex in-memory to later verify freshness
+    // of set-state request in a replica KME
+    state_req_nonce_hex = nonce_hex;
+
+    ByteArray nonce_ba = StrToByteArray(state_req_nonce_hex);
+    SetStatus(KME_REPL_OP_SUCCESS, out_work_order_data);
+    AddOutput(1, uid_ba, out_work_order_data); // unique_id_hex_str
+    AddOutput(2, nonce_ba, out_work_order_data); // generated_nonce_hex_str
+    ByteArray concat_ba = StrToByteArray( ByteArrayToStr(uid_ba)
+                                        + ByteArrayToStr(nonce_ba));
+    uint8_t concat_str_hash[SGX_HASH_SIZE] = {0};
+    ComputeSHA256Hash(ByteArrayToString(concat_ba), concat_str_hash);
+    EnclaveData* enclave_data = EnclaveData::getInstance();
+    ByteArray uid_nonce_signed = enclave_data->sign_message(
+        ByteArray(std::begin(concat_str_hash), std::end(concat_str_hash)));
+    uid_nonce_signed = StrToByteArray(ByteArrayToHexEncodedString(uid_nonce_signed));
+    AddOutput(3, uid_nonce_signed, out_work_order_data); // uid_nonce_signed
+}
+
+/*
+ *  Get Primary KME State API.
+ *
+ *  @param in_work_order_data - vector of work order indata
+ *  @param out_work_order_data - vector of work order outdata
+*/
+void KMEWorkloadProcessor::GetStateReplica(
+    const std::vector<tcf::WorkOrderData>& in_work_order_data,
+    std::vector<tcf::WorkOrderData>& out_work_order_data) {
+
+    // Extract attestation_info from in_data at index 0
+    ByteArray attestation_info = in_work_order_data[0].decrypted_data;
+    // Extract unique_id_hex from in_data at index 1
+    ByteArray unique_id_hex = in_work_order_data[1].decrypted_data;
+    // Extract generated_nonce_hex from in_data at index 2
+    ByteArray generated_nonce_hex = in_work_order_data[2].decrypted_data;
+    // Extract uid_nonce_sig from in_data at index 3
+    ByteArray uid_nonce_sig = in_work_order_data[3].decrypted_data;
+    // Extract replica KME pub signing key from in_data that comes in
+    // the request
+    ByteArray pub_sig_key_ba = in_work_order_data[4].decrypted_data;
```

Comment:
does attestation_info not contain public signing key and encryption key of the replica KME? Or is it only the attestation report?

Reply:
Have not included the attestation_info as of now. Can do away with the public keys once it is added.

Reply:
Alright